### DISCUSSION TITLE
Implementing a Reynolds- and Prandtl-dependent Heat Transfer Coefficient

### DISCUSSION THREAD

**Discussion by JKHINL** [idaholab/moose#18677] [Q&A General] [answered]
Hello,

I have a geometry consisting of prismatic blocks with holes cut out through them, such that fluid can flow through the holes. These blocks are heated and have a temperature of T_solid, which is defined as a global variable. Likewise, the fluid temperature, T_fluid, is also declared as a global variable, and is heated via convection with the blocks.

From hand calculations and equations of state, I know that the flow is laminar, but based on experimental data, I can't assume a uniform heat flux or surface temperature condition (which would make things much easier). I have a specific correlation that I would like to apply that is a function of the Reynolds and Prandtl numbers of the fluid for material property alpha (which is proportional to the heat transfer coefficient). 

I have had success implementing the Reynolds and Prandtl numbers as AuxVariables and AuxKernels, and can see these two variables in the postprocessor; however, when I try to implement these two into the GenericFunctionMaterial or ParsedMaterial -type Material blocks, Pronghorn informs me that it cannot recognize these two AuxVariables.

Is there a recommended way that I would be able to implement these two fluid properties into a material block? Alternatively, I do have the mass flux rho*v as a global variable, but I would need to be able to extract the absolute viscosity mu, and I am not quite sure how to do that from the FluidProperties block. Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Using a material property for alpha is the correct approach for this. 

GenericFunctionMaterial will not work will variables. The ParsedMaterial should work. If you paste that (or part of that) material definition and the variable definitions here we may be able to spot the problem. 

However, I would recommend you simply add a new material for your correlation. Besides, we would certainly welcome the contribution in Pronghorn. 

Another note is that the Reynolds and Prandtl numbers are already defined as material properties in PronghornFluidProps, so you could just add this material to have access to these. Note that if you are using automatic differentiation, auxiliary variables will not propagate the derivatives, so your Jacobian wont be as good with a material taking the AuxVariables as arguments. 

Do you have source code access on Pronghorn btw? If so we can point to you to specific examples of other materials.

Best,
Guillaume

**Reply by JKHINL** [reply]
Hi Guillaume,

I believe that I do have access to the source code, but I haven't made any contributions to it during my internship. Creating a separate material block specifically for this correlation that accepts the channel diameter, D, channel length, L, and fluid properties (Re, Pr, and fluid thermal conductivity, k) would certainly enrich my time here.

Here is my block of code so far (D and L would have to be hardcoded in this format, and k_fluid estimated to a single value):

```
  [Material_Properties_Greencast_Inner_Bypass]
    type = ParsedMaterial
    args = 'Re Pr'
    function = '${fparse 0.065*k_fluid*Re*Pr*(D/L)/(1+0.04*(Re*Pr*(D/L))^(2/3))}'
    material_property_names = 'alpha'
    block = '16'
  []
```

When it comes to .C files, I understand how to request user input for D and L. My biggest source of confusion is how the code knows how and where it could pull Re, Pr, and k_fluid.

**Reply by GiudGiud** [reply]
The ParsedMaterial does not make use of the fparse system. It s a different parser.
Please look at the example in [ParsedMaterial]() and follow this syntax instead. The list of parameters should help too
`material_property_names` is for the names of the material properties used in the `function` NOT for the name of the. material property you are declaring

If you make a new .C file instead of using the ParsedMaterial, then you can add variable parameters as
- params.addCoupledVar("var_name", "pronghorn default name", "var description")
But we would rather you used material properties,
so params.addParam<MaterialPropertyName>("mat prop name", "pronghorn namespace default name", "description").

Please see for example the KTADragCoefficient.C material for an example of retrieving material properties. 
And DragCoefficients.C for an example of declaring new material properties.

**Reply by JKHINL** [reply]
I am first trying to get ParsedMaterial to work because I ran into issues with compiling Pronghorn with HPC when I made some changes to which modules I accessed earlier this summer. If I have time before Thursday, I'll try to write my own .C and .h files.

I've been able to resolve some of the problems when it comes to ParsedMaterial. Here is the block I have so far:

  [Material_Properties_Greencast_Inner_Bypass]
    type = ParsedMaterial
    f_name = SiederTateHTC
    material_property_names = 'Re Pr k_fluid'
    function = '(k_fluid/0.01875^2) * (3.66+(0.065 * Re * Pr * (0.01875/2.0))/(1+0.04 * (Re * Pr * (0.01875/2.0))^(2/3)))'
    output_properties = alpha
    block = '16'

I get the following error in the log.out file:

*** ERROR ***
The requested regular material property Re is declared as an AD property. Either retrieve it as an AD property with getADMaterialProperty or declare it as a regular property with declareProperty[39m

I am not quite sure what this exactly means, or where I can go to do what it tells me to do.

**Reply by GiudGiud** [reply]
So there are two kinds of materials, the one using automatic differentiation (AD) to get a great Jacobian and the ones that are not (and have to have the derivatives implemented manually). 

Are you using PronghornFluidProps? This is using AD. Your parsed material is not.  You need to switch from ParsedMaterial to `ADParsedMaterial`

**Reply by JKHINL** [reply]
Works like a charm! Thank you very much!

My only remaining question pertains to the Reynolds number. When I look at it via the postprocessor, I see that it is substantially smaller than what I had handcalculated. The obvious reason would be that the porosity is factored in.

When I use Re for the ParsedMaterial module, should I correct it so that it does not take porosity into account? I am looking at the equations solved in Pronghorn:
 (https://hpcsc.hpc.inl.gov/ssl/PRONGHORN/site/input_manual/kernels.html) 

and I do not see an indication where alpha relies on the adjusted velocity (the velocity that results when implementing porous physics.)

**Reply by GiudGiud** [reply]
you can select between the Reynolds and the superficial reynolds number. Have a look at PronghornFluidProps.C to see more about this.
